torch/_dynamo/utils.py

```diff
@@ -1528,8 +1528,10 @@ def tensor_always_has_static_shape(
     Returns a tuple, where the first element is the bool of whether or not this tensor should have a static shape.
     The second element is a TensorStaticReason, useful for passing to tensor_static_reason_to_message if needed.
     """
-    if type(tensor) is torch.nn.Parameter:
+    if type(tensor) is torch.nn.Parameter and not config.assume_parameters_shapes_static_by_default:
         return True, TensorStaticReason.PARAMETER
+    if type(tensor) is torch.nn.Parameter and config.assume_parameters_shapes_static_by_default:
+        return False, None
```

Comment:
Do I understand this correctly that you want to try the dynamic shape treatment with this flag by default for Tensors and Parameters and if this flag is set to false, then it switches back to Tensor -> dynamic, Parameter -> static?

Reply:
Not sure I understood your explanation, but here's what I mean:

```
def tensor_always_has_static_shape(
    tensor: Union[torch.Tensor, Any], is_tensor: bool, guard_source: "GuardSource"
) -> Tuple[bool, TensorStaticReason]:
    """
    Given a tensor, source, and is_tensor flag, determine if a shape should be static.

    Args:
    tensor - the real tensor to evaluate, parameters force a static shape.
    is_tensor - internal dynamo check, esentially "is_tensor": target_cls is TensorVariable,
    tensors not in a TensorVariable for whatever reason are forced static.

    Returns a tuple, where the first element is the bool of whether or not this tensor should have a static shape.
    The second element is a TensorStaticReason, useful for passing to tensor_static_reason_to_message if needed.
    """
    if not config.attempt_tensor_always_has_static_shape:  # NEW
        return False, None  # NEW
    if type(tensor) is torch.nn.Parameter:
        return True, TensorStaticReason.PARAMETER
    if not is_tensor:
        return True, TensorStaticReason.NOT_TENSOR
    if guard_source.is_nn_module():
        return True, TensorStaticReason.NN_MODULE_PROPERTY
    return False, None
```

Reply:
Yes, this is what I understood. However, I personally would invert the meaning and have a config `attempt_tensor_always_has_dynamic_shape`. 

This is because if use `attempt_tensor_always_has_static_shape` and it is set to True, then the function would still return `False, None` for a regular Tensor, which to me is somewhat counterintuitive. Whereas, if we use `attempt_tensor_always_has_dynamic_shape` and set it to true, then the function will always return 
`False, None`, 
but if it is set to False, then the function will return still 
`False, None` for the Tensor, 
but `True, TensorStaticReason.PARAMETER`, for the Parameter. 

However, if we do it like that, then the checks for a `nn.module` and `not is_tensor `are bypassed?

I gave it another try, but left the `attempt_tensor_always_has_dynamic_shape=False` per default, as otherwise tests fail and I am not aware of all the implications this might have. What is your take on this?

Reply:
Keeping the old default is good.

Naming truly is the hardest problem in CS.

The problem with the current name is that it is unclear what the precedence with `assume_static_by_default` is. Even if `attempt_tensor_always_has_dynamic_shape=True`, if `assume_static_by_default=True` you will still get static shapes until automatic dynamic kicks in. Inverting the boolean makes it more clear that when you set it to False, there is still opportunity for the other configs to apply. I'm ok with wibbling the name though; how about `force_parameter_static_shapes = True` and `force_nn_module_property_static_shapes = True` on a per conditional basis (I don't think there is any use case for `not is_tensor`.

Reply:
I understand. How about this then
```
def tensor_always_has_static_shape(
    tensor: Union[torch.Tensor, Any], is_tensor: bool, guard_source: "GuardSource"
) -> Tuple[bool, TensorStaticReason]:
    """
    Given a tensor, source, and is_tensor flag, determine if a shape should be static.

    Args:
    tensor - the real tensor to evaluate, parameters force a static shape.
    is_tensor - internal dynamo check, esentially "is_tensor": target_cls is TensorVariable,
    tensors not in a TensorVariable for whatever reason are forced static.

    Returns a tuple, where the first element is the bool of whether or not this tensor should have a static shape.
    The second element is a TensorStaticReason, useful for passing to tensor_static_reason_to_message if needed.
    """
    if guard_source.is_nn_module() and config.force_nn_module_property_static_shapes:
        return True, TensorStaticReason.NN_MODULE_PROPERTY
    if type(tensor) is torch.nn.Parameter and config.force_parameter_static_shapes:
        return True, TensorStaticReason.PARAMETER
    if not is_tensor:
        return True, TensorStaticReason.NOT_TENSOR
    return False, None
```

where the default is `force_nn_module_property_static_shapes=True` and `force_parameter_static_shapes=True`

Reply:
yeah, lgtm

Reply:
I updated the files accordingly